pkg/promotion/runner/builtin/yaml_merger.go

```diff
@@ -0,0 +1,135 @@
+	"github.com/akuity/kargo/pkg/x/promotion/runner/builtin"
+	"github.com/akuity/kargo/pkg/yaml"
+)
+
+const stepKindYAMLMerge = "yaml-merge"
+
+func init() {
+	promotion.RegisterStepRunner(
+		stepKindYAMLMerge,
+		promotion.StepRunnerRegistration{Factory: newYAMLMerger},
+	)
+}
+
+// yamlMerger is an implementation of the promotion.StepRunner interface that
+// merges multiple YAML files.
+type yamlMerger struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newYAMLMerger returns an implementation of the promotion.StepRunner interface
+// that merges multiple YAML files.
+func newYAMLMerger(promotion.StepRunnerCapabilities) promotion.StepRunner {
+	return &yamlMerger{schemaLoader: getConfigSchemaLoader(stepKindYAMLMerge)}
+}
+
+// Run implements the promotion.StepRunner interface.
+func (y *yamlMerger) Run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+) (promotion.StepResult, error) {
+	cfg, err := y.convert(stepCtx.Config)
+	if err != nil {
+		return promotion.StepResult{
+			Status: kargoapi.PromotionStepStatusFailed,
+		}, &promotion.TerminalError{Err: err}
+	}
+	return y.run(ctx, stepCtx, cfg)
+}
+
+// convert validates yamlMerger configuration against a JSON schema and
+// converts it into a builtin.YAMLMergeConfig struct.
+func (y *yamlMerger) convert(cfg promotion.Config) (builtin.YAMLMergeConfig, error) {
+	return validateAndConvert[builtin.YAMLMergeConfig](y.schemaLoader, cfg, stepKindYAMLMerge)
+}
+
+func (y *yamlMerger) run(
+	_ context.Context,
+	stepCtx *promotion.StepContext,
+	cfg builtin.YAMLMergeConfig,
+) (promotion.StepResult, error) {
+	// Validate and collect input file paths
+	filePaths := make([]string, 0, len(cfg.InFiles))
+	for _, path := range cfg.InFiles {
+		inFile, err := securejoin.SecureJoin(stepCtx.WorkDir, path)
+		if err != nil {
+			return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+				fmt.Errorf("invalid input file path %q: %w", path, err)
+		}
+
+		if _, err = os.Stat(inFile); err != nil {
+			if os.IsNotExist(err) && cfg.IgnoreMissingFiles {
+				continue
+			}
+			return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+				fmt.Errorf("input file %q not found: %w", path, err)
+		}
+		filePaths = append(filePaths, inFile)
+	}
+
+	// Require at least one input file to merge
+	if len(filePaths) == 0 {
+		return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+			fmt.Errorf("no input files found to merge")
+	}
+
+	// Validate output path
+	outFile, err := securejoin.SecureJoin(stepCtx.WorkDir, cfg.OutFile)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+			fmt.Errorf("invalid output file path %q: %w", cfg.OutFile, err)
+	}
+
+	// Ensure output directory exists
+	if err = os.MkdirAll(filepath.Dir(outFile), 0o700); err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+			fmt.Errorf("error creating directory %q: %w", filepath.Dir(outFile), err)
```

Comment:
Nit: Since the directory very well may exist already, I would have this say there was an error "ensuring the existence" of the directory instead of saying it was an error creating it.